### PR TITLE
@GenerateMapping: bean-shaped wire targets (#628)

### LINKED ISSUE #628: [FEAT] @GenerateMapping: bean-shaped targets — map getter/setter DTOs, not just records
> **Design decisions locked** (review, 2026-07-18): the bean contract and every open question below are resolved in [this comment](https://github.com/higher-kinded-j/higher-kinded-j/issues/628#issuecomment-5012882508) — wire-side beans only; detection ladder = JavaBeans setters → builder (`builder()`/`newBuilder()`) → JAXB collection-getter, all-args constructors deferred; null-guarded parse (located `FieldError`s); `Optional<T>` ↔ nullable `T` bridging in scope; projections degrade to a **validated patch** instead of `asLens()`; sparse null-as-absent PATCH (`Update<Domain>`) split out as a follow-up.

## The gap

`@GenerateMapping` ([#600](https://github.com/higher-kinded-j/higher-kinded-j/issues/600), Step-0 [#605](https://github.com/higher-kinded-j/higher-kinded-j/pull/605), remainder [#621](https://github.com/higher-kinded-j/higher-kinded-j/pull/621)) maps **records and sealed hierarchies** — it reads components positionally and constructs via the canonical constructor. The one un-owned shape it cannot yet target is the **bean-shaped DTO**: a mutable class with a no-args constructor and getters/setters, or an immutable one with a builder. These dominate generated wire types (JAXB, protobuf-lite, many OpenAPI generators), so their absence is the main thing blocking "map the DTO you were handed."

## What "good" looks like (user perspective)

```java
public record User(String name, EmailAddress email) {}

// A generated, mutable, getter/setter bean - not annotatable, not a record.
public class UserDto { String getName(); void setName(String); String getEmail(); void setEmail(String); }

@GenerateMapping
public interface UserBeanMapping extends MappingSpec<User, UserDto> {}
// build(User)     -> new UserDto(); setName(...); setEmail(...);
// parse(UserDto)  -> read via getName()/getEmail(), located + accumulating as usual
```

- **Bean targets are read via getters and written via setters/builder**, chosen by the bean's actual shape.
- **The annotation still sits on the spec**, never on the (un-owned) bean — third-party beans map without being annotatable, same as records today.
- **All existing features apply** — renames, leaves, nesting, container lifting, emission tiers — with the bean shape only changing *how* components are read and written.

## Things to explore / constraints

- **Property model.** A bean-property analyser (getter/setter/builder discovery) alongside the record-component path; define name/type matching over properties.
- **Construction strategy.** No-args + setters, all-args constructor, or builder — detect and pick, with a diagnostic when none is usable.
- **Emission-tier truthfulness.** A bean with no faithful inverse (write-only or partial) must degrade tiers honestly (no fabricated `asIso`).
- **Nullability & defaults** of bean properties vs record components.
- **Diagnostics** for un-constructable / unreadable beans, what/why/fix.

## Dependencies / scope

- **Depends on** #605 and the remainder (#621) — both merged, so this is unblocked.
- **In:** record/sealed ↔ bean mappings (getters/setters/builder), reusing the full feature set with a bean property model.
- **Out:** reflective runtime mapping (stays compile-time); non-JavaBeans conventions unless trivially covered; all-args-constructor beans (compiled third-party classes lack reliable parameter names); sealed dispatch over bean subtypes (un-owned hierarchies aren't sealed); sparse null-as-absent PATCH write-back (`Update<Domain>` from non-null fields) — explicit opt-in, tracked in #645.

---

**Implemented in PR #644** (slices A–F): the `WireShape` abstraction, mutable-JavaBean/builder/JAXB construction strategies, null-guarded parse, and `Optional<T>` bridging. Deferred to follow-ons (documented in the book and `docs/plans/IMPL-PLAN-bean-mapper.md`): the validated-`patch` projection tier for reference-typed bean projections (all-primitive projections get `asLens()` today), read-only/write-only bean tiers, and Optional bridging through a nested spec. The sparse null-as-absent variant is #645.

### PR DESCRIPTION
**Pull request by MagnusSmith**
Closes #628.

Adds **bean-shaped wire targets** to `@GenerateMapping`: a record domain can now map to a bean (getter/setter DTO or builder-based immutable), not only a record. This is the one un-owned wire shape the mapper couldn't reach, and it dominates generated DTOs (JAXB, protobuf-lite, many OpenAPI generators). Design decisions were locked in the [#628 review comment](https://github.com/higher-kinded-j/higher-kinded-j/issues/628#issuecomment-5012882508); the plan lives at `docs/plans/IMPL-PLAN-bean-mapper.md`.

Zero new public API — the same empty spec interface, with the wire shape detected:

```java
public record User(String name, EmailAddress email) {}
public class UserDto { String getName(); void setName(String); String getEmail(); void setEmail(String); }

@GenerateMapping
public interface UserMapping extends MappingSpec<User, UserDto> {
  default ValidatedPrism<String, EmailAddress> email() { return EmailCodecs.EMAIL; }
}
// build(User)    -> new UserDto(); setName(...); setEmail(...)
// parse(UserDto) -> reads getName()/getEmail(), null-guarded, located + accumulating
```

## What's included

- **Construction ladder** (detected from the bean's shape): a no-args constructor with `setX` setters (and, for a getter-only `List`, the JAXB convention `getItems().addAll(...)`); then a static `builder()`/`newBuilder()` whose setters fill it and whose `build()` yields the wire. Fluent setters (returning the bean) are accepted.
- **Null is located, never thrown.** An unset bean property is `null` and `ValidatedPrism.parse` rejects `null`, so every reference-typed read is null-guarded into a located `FieldError`; `parse` stays total and accumulating, and a `null` never reaches a leaf. Primitive reads are unguarded.
- **Honest tiers, for free.** Those guards make reference components fallible, so `asIso()` is withheld automatically — an all-primitive bean still earns it. Bean mappings expose `asValidatedPrism()`, so record specs nest them and containers lift them.
- **Optional bridging.** A domain `Optional<T>` maps to a nullable bean property `T` (empty ↔ absent): `build` skips the write when empty (protecting null-hostile setters), `parse` reads `Optional.ofNullable(...)`.
- All existing features carry over: renames, leaves, derived fields, `List`/`Optional`/`Map` lifting, nesting both ways, sealed dispatch.
- What/why/fix diagnostics for: a bean domain, an unusable bean, a getter/setter type mismatch, a non-record/non-bean wire, and a deferred reference-typed bean projection.

## Structure (sequential commits)

- **A** — a `WireShape` abstraction over the wire side (zero behaviour change; byte-identical record output).
- **B** — mutable JavaBean full tier + null-guarded parse.
- **C** — builder and JAXB collection strategies.
- **D** — Optional bridging and the bean-projection safety gate.
- **E** — book section, example, law-check, javadoc, release notes.

## Deferred to follow-ons (documented)

The validated-`patch` projection tier (reference-typed bean projections), read-only (parse-only) and write-only (build-only) bean tiers, and Optional bridging through a nested spec. An all-primitive bean projection maps as a lawful `asLens()` today; a reference-typed one is reported with a pointer to the `patch` tier rather than emitting an unlawful lens. The sparse null-as-absent PATCH follow-up (`Update<Domain>`) is drafted at `docs/plans/DRAFT-ISSUE-bean-sparse-update.md`.

## Gates

`MappingProcessor*` holds at 100% line and branch coverage; the analyser and `WireShape` stay above the package gate; the full project builds clean; book snippets are compile-verified in hkj-examples; every emission tier (including the bean tier) is law-checked through the published `MappingLaws` harness.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for mapping record domains to bean-shaped wire types (mutable getter/setter DTOs and immutable builder-based DTOs), including JavaBeans `isX` properties and common `List` getter/add patterns.
  * Extended `Optional<T>` mappings by bridging to nullable bean properties while keeping `parse` total via located `FieldError`s for unset reference fields.
* **Documentation**
  * Updated mapping docs, release history, and examples to cover bean-shaped DTOs and correct annotation placement.
* **Tests**
  * Added/expanded processor, law, and example tests for multiple bean construction strategies and null/Optional edge cases.

### CODE REVIEW

File: hkj-book/src/optics/record_mapping.md [hunk 1]
```diff
@@ -190,7 +190,33 @@ The overloads follow the tiers:
...
+
+The wire side need not be a record. A **bean** (a mutable class with a no-args constructor and getters/setters, or an immutable one with a builder) maps the same way, with the same features (renames, leaves, derived fields, container lifting, nesting). Only *how* the wire is read and written changes: `build` fills through setters or a builder, and `parse` reads through getters.
+
+``` java
+{{#include ../../../hkj-examples/src/main/java/org/higherkindedj/example/book/mapping/RecordMappingBook.java:bean_spec}}
+```
+
+``` java
+{{#include ../../../hkj-examples/src/main/java/org/higherkindedj/example/book/mapping/RecordMappingBook.java:bean_usage}}
+```
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply hkj-book/src/optics/record_mapping.md:208]: This is a false positive — leaving these as backtick fences.

- **No markdownlint in this repo.** There's no `.markdownlint*` config, and nothing runs markdownlint in CI (`.github/workflows`, `package.json`, `book.toml` all clean). There's no rule to satisfy here.
- **Backtick is the book-wide convention for code.** Across `hkj-book/src` there are ~5,888 top-level ` ``` ` code fences; `~~~` is used *only* to open the ~1,223 mdbook-admonish containers. These two blocks use backticks, exactly like every other code block in the book.
- Switching them to `~~~` would make them the **only** tilde code fences in the entire book — which is the actual MD048 inconsistency. (The comment's own reasoning agrees: the `~~~` exception is for nested fences *inside* admonish containers; these are ordinary top-level fences, so they belong on backticks.)

File: .claude/skills/hkj-mapping/reference/mapping-example.md [hunk 2]
```diff
@@ -144,6 +144,47 @@ public interface PersonMapping extends MappingSpec<Person, PersonDto> {
...
+        EmailAddress::value);
+  }
+}
+```
+
+A bean property is `null` when unset, and a leaf's `parse` rejects `null`, so every reference read is
+null-guarded: a missing field becomes a located `FieldError` (`email: must not be null`) rather than
+throwing, and `parse` stays total and accumulating. An immutable bean with a static `builder()` or
+`newBuilder()` (Lombok, Immutables, AutoValue, protobuf) maps identically - `build` goes through the
+builder. A domain `Optional<T>` bridges to a nullable bean property `T` (empty maps to absent).
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply .claude/skills/hkj-mapping/reference/mapping-example.md]: Fixed in 6078033d8. Reworded so the ordering is explicit: the mapper null-guards each nullable getter result *before* handing it to the leaf, so a missing field becomes a located `FieldError` (`email: must not be null`) rather than an NPE — `parse` stays total and accumulating.

File: .claude/skills/hkj-mapping/SKILL.md [hunk 3]
```diff
@@ -188,6 +188,10 @@ rejected outright: the projection's `asLens()` write-back could never honour a c
   key.
 - **The mapped record need not be yours.** The annotation sits on *your spec interface*, never on
   the record, so third-party and library records map fine.
+- **The wire need not be a record.** A bean-shaped DTO (getters/setters, or an immutable one with a
+  static `builder()`/`newBuilder()`) maps too - `build` fills through setters or the builder, `parse`
+  reads through getters. A `null` bean property becomes a located `FieldError` (never thrown), and a
+  domain `Optional<T>` bridges to a nullable bean property `T`. See `reference/mapping-example.md`.
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply .claude/skills/hkj-mapping/SKILL.md]: Good catch — added. Both `SKILL.md` and `reference/mapping-example.md` now list the JAXB collection-getter convention (a getter-only `List` whose getter returns a live mutable list, filled with `getItems().addAll(...)`) as the third detected construction shape, alongside a no-args constructor with setters and a static `builder()`/`newBuilder()`. Fixed in 6078033d8.

File: hkj-examples/src/main/java/org/higherkindedj/example/optics/GenerateMappingExample.java [hunk 4]
```diff
@@ -290,7 +377,35 @@ public static void main(String[] args) {
...
+            + " as must-not-be-null (a null never reaches the leaf)\n");
+
+    System.out.println("=== Builder-Based Bean Example (immutable DTO via builder()) ===");
+    Point origin = new Point(3, "origin");
+    PointDto pointDto = GenerateMappingExamplePointMappingImpl.INSTANCE.build(origin);
+    System.out.println(
+        "Built:        PointDto[x=" + pointDto.getX() + ", label=" + pointDto.getLabel() + "]");
+    System.out.println(
+        "Round trip:   " + GenerateMappingExamplePointMappingImpl.INSTANCE.parse(pointDto));
+    System.out.println("Expected: built through Point.builder(); the round trip is Valid");
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply hkj-examples/src/main/java/org/higherkindedj/example/optics/GenerateMappingExample.java]: Fixed in 6078033d8. `Point` is the domain record; the generated mapping builds through `PointDto.builder()` (the DTO's builder), not `Point.builder()`. The example output now names `PointDto.builder()`.
